### DISCUSSION TITLE
alert threshold, names in static texts, weather support, Atlantis theme

### DISCUSSION THREAD

**Discussion by GaryStar** [mathoudebine/turing-smart-screen-python#429] [Ideas]
@mathoudebine: I've added several features to the base code and I've created a fork to store the branch with my changes. If you like, please check it and see it is something that can be integrated in your code.
https://github.com/GaryStar/turing-smart-screen-python/tree/feature/names-alert-weather
Features added:
 - ALERT_VALUE and ALERT_COLOR keys in theme. You can specify a threshold value in the ALERT_VALUE key and if the monitored value will be greater or equal to that, the color of the element (TEXT, GRAPH or RADIAL) will be changed to ALERT_COLOR (or 255, 0, 0 if not specified)
 - detected names of CPU, GPU and Memory can be used in static_text elements (value for TEXT key) as names.Cpu, names.Gpu or names.Memory
 - weather support. OpenWeatherMap API key is required. Can be obtained by creating a free account on https://home.openweathermap.org/users/sign_up . City, API Key and Units can be configured with configure.py script
 - memory values for GPU and RAM will auto scale (change unit) based on size. Ex: 34287923200 bytes will be displayed as 31.9 G
 - Redraw background on wake from sleep can be enabled or disabled from configure.py script
 - Added Atlantis theme (layout based on Cyberpunk theme, changed graphics) that will make use of all the above features.
 
![Screenshot 2023-12-22 182613](https://github.com/mathoudebine/turing-smart-screen-python/assets/7147048/59605821-3aab-46ca-9459-a405e8356a01)

**Comment by GaryStar** [comment]
Update:
- added ICON option for weather
- added 2 sets of weather icons (a default one for all themes and one for Atlantis theme)
- added function to display PNG images with transparency and scale
Example usage:
```
  WEATHER:
    INTERVAL: 600
    TEXT:
      SHOW: True
      X: 30
      Y: 52
      FONT: roboto-mono/RobotoMono-Bold.ttf
      FONT_SIZE: 18
      FONT_COLOR: 150, 200, 200
      BACKGROUND_IMAGE: background.png
    ICON:
      SHOW: True
      X: 30
      Y: 11
      SCALE: 0.43
      BACKGROUND_IMAGE: background.png
```
![preview](https://github.com/mathoudebine/turing-smart-screen-python/assets/7147048/27077c2b-9c48-4c2b-ac94-74d9df9758b0)
default weather icons (will be used for any theme that don't have its own icons)
![default_weather_icons](https://github.com/mathoudebine/turing-smart-screen-python/assets/7147048/be121704-b5a7-4b57-8741-8f67b6f3e331)
Specific icons for a theme (will be used instead of default icons)
![atlantis_weather_icons](https://github.com/mathoudebine/turing-smart-screen-python/assets/7147048/38dc0293-68ea-40d5-b417-aeee27bdb81d)

**Comment by mathoudebine** [comment]
Thanks for sharing your work here, I will definitely check your fork! The weather monitoring might be a good addition to this project 👍

**Comment by AlexDroid00** [comment]
I really like your fork. Good work! I use it daily 👍

**Comment by DJRoby19** [comment]
Hi I'm trying this fork but it doesn't boot on linux mint from this error

root@Ryzen:/opt/turing-smart-screen-python-feature-names-alert-weather# python3 main.py
14/02/2024 00:45:41 [INFO] Loading theme Atlantis from res/themes/Atlantis/theme.yaml
14/02/2024 00:45:41 [DEBUG] HW revision: A
14/02/2024 00:45:41 [DEBUG] Auto detected COM port: /dev/ttyACM0
Traceback (most recent call last):
  File "/opt/turing-smart-screen-python-feature-names-alert-weather/main.py", line 78, in <module>
    class POWERBROADCAST_SETTING(Structure):
NameError: name 'Structure' is not defined
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 153, in apport_excepthook
    with os.fdopen(os.open(pr_filename,
FileNotFoundError: [Errno 2] File o directory non esistente: '/var/crash/_opt_turing-smart-screen-python-feature-names-alert-weather_main.py.0.crash'

Original exception was:
Traceback (most recent call last):
  File "/opt/turing-smart-screen-python-feature-names-alert-weather/main.py", line 78, in <module>
    class POWERBROADCAST_SETTING(Structure):
NameError: name 'Structure' is not defined
root@Ryzen:/opt/turing-smart-screen-python-feature-names-alert-weather#

**Reply by GaryStar** [reply]
Sorry, my mistake. That class should be defined only for Windows platform.
please add ```if platform.system() == "Windows":``` before that line (78) and indent the next 4 lines (that small class).
Like this:
```
if platform.system() == "Windows":
    class POWERBROADCAST_SETTING(Structure):
        _fields_ = [("PowerSetting", GUID),
                    ("DataLength", DWORD),
                    ("Data", DWORD)]
```
or just move that class between lines 142 and 143:
```
    if platform.system() == "Windows":
    # ---> move the class there
        def on_win32_ctrl_event(event):
```

**Comment by DJRoby19** [comment]
Ok now it works
we need to fix the graphics because the weather icon is too big and the writing is off the screen, I'll fix it now thanks

**Reply by GaryStar** [reply]
for weather icon you have "SCALE" parameter so you can adjust it as you like:
Ex:
```
    ICON:
      SHOW: True
      X: 30
      Y: 11
      SCALE: 0.43
      BACKGROUND_IMAGE: background.png
```
